clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java

```diff
@@ -194,27 +218,38 @@ public void testSuccessfulHeartbeatTiming() {
             heartbeatRequestState.timeToNextHeartbeatMs(time.milliseconds()),
             "Time to next interval was not properly updated.");
 
+        t = time.milliseconds();
+        when(heartbeatRequestState.canSendRequest(t)).thenReturn(false);
+        t = t + DEFAULT_HEARTBEAT_INTERVAL_MS - partOfInterval;
+        when(heartbeatRequestState.canSendRequest(t)).thenReturn(true);
         inflightReq.handler().onComplete(createHeartbeatResponse(inflightReq, Errors.NONE));
         assertNextHeartbeatTiming(DEFAULT_HEARTBEAT_INTERVAL_MS - partOfInterval);
     }
 
-    @ParameterizedTest
+    @Test
     @ApiKeyVersionsSource(apiKey = ApiKeys.CONSUMER_GROUP_HEARTBEAT)
-    public void testFirstHeartbeatIncludesRequiredInfoToJoinGroupAndGetAssignments(short version) {
-        resetWithZeroHeartbeatInterval(Optional.of(DEFAULT_GROUP_INSTANCE_ID));
+    public void testFirstHeartbeatIncludesRequiredInfoToJoinGroupAndGetAssignments() {
+        resetWithZeroHeartbeatInterval();
         String topic = "topic1";
         subscriptions.subscribe(Collections.singleton(topic), Optional.empty());
         membershipManager.onSubscriptionUpdated();
 
         // Create a ConsumerHeartbeatRequest and verify the payload
         assertEquals(0, heartbeatRequestManager.maximumTimeToWait(time.milliseconds()));
+        when(heartbeatRequestState.canSendRequest(anyLong())).thenReturn(true);
         NetworkClientDelegate.PollResult pollResult = heartbeatRequestManager.poll(time.milliseconds());
         assertEquals(1, pollResult.unsentRequests.size());
         NetworkClientDelegate.UnsentRequest request = pollResult.unsentRequests.get(0);
         assertInstanceOf(Builder.class, request.requestBuilder());
 
-        ConsumerGroupHeartbeatRequest heartbeatRequest =
-                (ConsumerGroupHeartbeatRequest) request.requestBuilder().build(version);
+        ConsumerGroupHeartbeatRequest heartbeatRequest = mock(ConsumerGroupHeartbeatRequest.class);
+        ConsumerGroupHeartbeatRequestData data = new ConsumerGroupHeartbeatRequestData();
+        data.setSubscribedTopicNames(Collections.singletonList(topic));
+        data.setRebalanceTimeoutMs(10000);
+        data.setGroupId("groupId");
+        data.setInstanceId("group-instance-id");
+
+        when(heartbeatRequest.data()).thenReturn(data);
```

Comment:
uhm we shouldn't create the request data manually. We do want to test that the HBManager builds that data correctly from the state it has (the [buildRequestData](https://github.com/apache/kafka/blob/7d702e17604c6cc95e4e8bfdea58b04be3f5ffc3/clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java#L551) func inside the `heartbeatState`). So we should have the state as an actual instance and have the test as it was before I expect.